### PR TITLE
[Glitch] [Proposal] Make able to write React in Typescript

### PR DESCRIPTION
**Pull request by Plastikmensch**
Port 4520e6473afe901005e7ac532317f4b4f1af9ead to glitch-soc

Instead of adding ts-expect-error comments to initial_state, I added the properties to the typedefs.
Refactored `app/javascript/flavours/glitch/components/avatar.tsx` similar to 9f8d34620b6fa502da683f3775805bf2a5474f92
Needed to adjust tsconfig.json to also include glitch-soc files and also added paths so `'flavours/glitch/...'` imports work.

### CODE REVIEW

File: app/javascript/flavours/glitch/components/avatar.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import * as React from 'react';
+import classNames from 'classnames';
+import { autoPlayGif } from '../initial_state';
+import { useHovering } from '../../../hooks/useHovering';
+import type { Account } from '../../../types/resources';
```
Comment (Plastikmensch) [inline app/javascript/flavours/glitch/components/avatar.tsx:5]: I really don't like doing imports like this, but changing tsconfig.json to allow 'flavours/glitch/...' type imports feels out of scope for this PR and also makes some of the ts-expect-error comments obsolete.

File: app/javascript/flavours/glitch/components/avatar.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import * as React from 'react';
+import classNames from 'classnames';
+import { autoPlayGif } from 'flavours/glitch/initial_state';
+import { useHovering } from 'hooks/useHovering';
+import type { Account } from 'types/resources';
+
+type Props = {
+  account: Account;
+  className: string;
+  size: number;
```
Comment (Plastikmensch) [inline app/javascript/flavours/glitch/components/avatar.tsx:10]: Only `size` is required.
`account` is treated as required, but can be undefined if an account is suspended.

File: tsconfig.json [hunk 3]
```diff
@@ -7,7 +7,15 @@
     "noEmit": true,
     "strict": true,
     "esModuleInterop": true,
-    "skipLibCheck": true
+    "skipLibCheck": true,
+    "baseUrl": ".",
+    "paths": {
+      "*": ["*", "app/javascript/*"]
+    }
```
Comment (ClearlyClaire) [inline tsconfig.json:14]: What are these changes?

File: app/javascript/flavours/glitch/initial_state.js [hunk 4]
```diff
@@ -81,14 +82,19 @@
  * @property {boolean=} use_pending_items
  * @property {string} version
  * @property {boolean} translation_enabled
- * @property {object} local_settings
```
Comment (ClearlyClaire) [inline app/javascript/flavours/glitch/initial_state.js:84]: Why did you remove `local_settings`?

File: app/javascript/flavours/glitch/initial_state.js [hunk 5]
```diff
@@ -81,14 +82,19 @@
...
+ * @property {string} status_page_url
```
Comment (ClearlyClaire) [inline app/javascript/flavours/glitch/initial_state.js:85]: `status_page_url` should be an upstream thing… I'll investigate.

File: app/javascript/flavours/glitch/initial_state.js (hunk 4 above)
Comment (Plastikmensch) [reply app/javascript/flavours/glitch/initial_state.js:84]: It was applied to the wrong type def.
I moved it to the typedef of InitialState

File: app/javascript/flavours/glitch/initial_state.js (hunk 5 above)
Comment (Plastikmensch) [reply app/javascript/flavours/glitch/initial_state.js:85]: It is.
Upstream didn't add it to the typedef.
I can replace it with an expect error comment, if you want to.

File: tsconfig.json (hunk 3 above)
Comment (ClearlyClaire) [reply tsconfig.json:14]: Can you make that change?

Comment (ClearlyClaire) [reply tsconfig.json:14]: Also why would that be needed for glitch-soc, but not upstream Mastodon for `import * from 'mastodon/...'`?

**Review by ClearlyClaire — APPROVED** [review]

**Comment by Plastikmensch** [comment]
> One thing I missed when reviewing this is that it's using common files in `app/javascript/types`, which is going to be an issue. I also completely missed [e38b391](https://github.com/glitch-soc/mastodon/commit/e38b39194075bec88af2f3a5134137563024d391) for some reason.

You mean that `app/javascript/types` is used both in vanilla and glitch-soc?
Yeah, it slipped my mind that this would cause issues.
Sorry about that.
